### PR TITLE
Handle Any formatting for diagnostic purposes

### PR DESCRIPTION
**Pull request by jskeet**
See #1008 for background.

### CODE REVIEW

**Comment by jskeet** [comment]
// cc @anandolee @xfxyjwf

File: csharp/src/Google.Protobuf/JsonFormatter.cs [hunk 1]
```diff
@@ -516,6 +541,12 @@ private void WriteFieldMask(StringBuilder builder, IMessage value)
...
         {
+            if (ReferenceEquals(this, diagnosticFormatter))
+            {
+                WriteDiagnosticOnlyAny(builder, value);
+                return;
+            }
+
             string typeUrl = (string) value.Descriptor.Fields[Any.TypeUrlFieldNumber].Accessor.GetValue(value);
             ByteString data = (ByteString) value.Descriptor.Fields[Any.ValueFieldNumber].Accessor.GetValue(value);
             string typeName = GetTypeName(typeUrl);
```
Comment (anandolee) [inline csharp/src/Google.Protobuf/JsonFormatter.cs]: Do we have any or if it is possible to add tests for Any message when use TypeRegistry?

Comment (jskeet) [inline csharp/src/Google.Protobuf/JsonFormatter.cs]: Yup, quite a few in JsonFormatter and JsonParser. (Although if you want to suggest any more, I'm sure there's always room!)

**Comment by jtattermusch** [comment]
Code-wise, this looks alright. I am still a bit upset about #1008, but this approach sounds like a reasonable solution for the problem.

From user's point of view, it's unfortunate using message.ToString() becomes a bit of an antipattern from now on. But perhaps making user clearly distinguish between ToString() and json serialization is a good thing in general. 

From C# code point of view, this is LGTM, but we should absolutely make sure that the approach we take here is consistent with what Java (and other languages that have ToString equivalent) is doing.  Leaving the PR to @xfxyjwf  and @anandolee  to make the call.

File: csharp/src/Google.Protobuf/JsonFormatter.cs (hunk 1 above)
Comment (anandolee) [inline csharp/src/Google.Protobuf/JsonFormatter.cs]: Oh I've found the WKT in Any and nested Any. I think that's good enough.

LGTM

**Comment by xfxyjwf** [comment]
On Tue, Dec 15, 2015 at 4:14 PM, Jan Tattermusch <notifications@github.com>
wrote:

> Code-wise, this looks alright. I am still a bit upset about #1008
> <https://github.com/google/protobuf/issues/1008>, but this approach
> sounds like a reasonable solution for the problem.
>
> From user's point of view, it's unfortunate using message.ToString()
> becomes a bit of an antipattern from now on. But perhaps making user
> clearly distinguish between ToString() and json serialization is a good
> thing in general.
>
> From C# code point of view, this is LGTM, but we should absolutely make
> sure that the approach we take here is consistent with what Java (and other
> languages that have ToString equivalent) is doing. Leaving the PR to
> @xfxyjwf <https://github.com/xfxyjwf> and @anandolee
> <https://github.com/anandolee> to make the call.
>
The ToString() implementations in other languages like C++ and Java use
text format instead of JSON format. Users are advised to avoid using it as
a wire protocol because the output of text format is not necessarily
parsable by other languages and not even by itself if the message contains
unknown fields. Drawing a line between when to use ToString() (for
debugging/logging purpose) and when to use the json serialization (as a
wire protocol) seems good to me.

**Comment by jtattermusch** [comment]
@xfxyjwf thanks for explanation, that totally makes sense. LGTM.

**Comment by jskeet** [comment]
Given the LGs above, merging.

**Pull request merged**